test/Parse/matching_patterns.swift

```diff
@@ -289,22 +289,44 @@ case (let (_, _, _)) + 1:
   ()
 }
 
-// FIXME: We don't currently allow subpatterns for "isa" patterns that
-// require interesting conditional downcasts.
-class Base { }
-class Derived : Base { }
+// "isa" patterns.
 
+// https://github.com/apple/swift/issues/56139
+// Allow subpatterns for "isa" patterns that require conditional
+// collection downcasts
+do {
+  class Base { }
+  class Derived : Base { }
 
-switch [Derived(), Derived(), Base()] {
-case let ds as [Derived]: // expected-error{{collection downcast in cast pattern is not implemented; use an explicit downcast to '[Derived]' instead}}
-  ()
-case is [Derived]: // expected-error{{collection downcast in cast pattern is not implemented; use an explicit downcast to '[Derived]' instead}}
-  ()
+  var arr: [Base]
 
-default:
-  ()
-}
+  if case let _ as [Derived] = arr {}
+
+  guard case let _ as [Derived] = arr else {}
+
+  while case let _ as [Derived] = arr {}
+
```

Comment:
Please add analogous tests for the `is` pattern, and one more test featuring a for-in loop for each pattern. Example: `for case _ as [Derived] in [arr]` (note that we are casting the *element* here, and it must be a collection downcast).